src/net/thread.rs

```diff
@@ -0,0 +1,85 @@
+use std::sync::{mpsc, Arc, Mutex};
+use std::thread;
+
+pub struct Thread {
+    worker: Worker,
+    sender: mpsc::Sender<Message>,
+}
+
+pub struct Worker {
+    id: usize,
+    thread: Option<thread::JoinHandle<()>>,
```

Comment:
I was thinking of a case where I terminate the thread, it'll perform `take()` and then `join()` which will leave None in `thread` indicating that this worker has no running thread. I thought Option would be a good fit, but I'm pretty new to Rust so I'd be happy to learn if there's a better way to do it :)

Reply:
The best way to manage that is to have the function that terminates the thread take `self` (not `&self`) so that the object is consumed.

Reply:
I was using `Drop` which takes `&mut self` to terminate the thread, instead of creating a function within struct. Would creating a function that takes `self` to terminate the thread better than using `drop()`?

Reply:
I'll note that right now the Drop implementation ensures that you have to drop the object in order for it to stop being useful.  That has the same effect as a function that consumes the object.

Reply:
I was thinking of a case where I want to reuse a `worker.thread`; if a thread is terminated, spawn a new thread. But again, that's probably more relevant in thread pool case, and I can achieve the same by removing the dead worker and adding a new one.

So, just to make sure I understand correctly, do you suggest not using `Drop` trait and create a separate function (ie. `shutdown(self)`) to `join`, or still use `Drop` as well?

If I implement `Drop`, the following would fail:

```
impl Thread {

  pub fn shutdown(self) {
    self.worker.thread.join().unwrap(); // <- cannot move out of Thread because JoinHandle doesn't implement Copy
  }

impl Drop for Thread {
  fn drop(&mut self) {
```

Since `join(self)` consumes the object, the same error occurs in `drop(&mut self)`. One way to mitigate is to not implement `Drop` but I didn't think that's a good option.

`Option` handles this case nicely becase we can `take`. Is there anyway we can move an object behind a mutable reference?

Reply:
OK, now that I've looked into this more, the Option seems to be the only way that you can have a Drop implementation that consumes a field on a struct.  This is what the internal JoinInner implementation does for JoinHandle, for instance.

The alternative is to not implement Drop, but that means that we potentially leak threads if something isn't cleaned up properly... and manually.  I don't like manual cleanup.